test/validator/strict_dependencies_test.dart

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:path/path.dart' as path;
+import 'package:pub/src/entrypoint.dart';
+import 'package:pub/src/validator.dart';
+import 'package:pub/src/validator/strict_dependencies.dart';
+import 'package:scheduled_test/scheduled_test.dart';
+
+import '../descriptor.dart' as d;
+import '../test_pub.dart';
+import 'utils.dart';
+
+Validator strictDeps(Entrypoint entrypoint) {
+  return new StrictDependenciesValidator(entrypoint);
+}
+
+main() {
+  group('should consider a package valid if', () {
+    setUp(d.validPackage.create);
+
+    integration('looks normal', () => expectNoValidationError(strictDeps));
+
+    integration('declares an "import" as a dependency', () {
+      d.dir(appPath, [
+        d.libPubspec("test_pkg", "1.0.0", deps: {
+          "not_declared": "^1.2.3"
+        }, sdk: ">=1.8.0 <2.0.0")
+      ]).create();
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import 'package:not_declared/not_declared.dart';
+      ''').create();
+      expectNoValidationError(strictDeps);
+    });
+  });
+
+  group('should consider a package invalid if', () {
+    setUp(d.validPackage.create);
+
+    integration('does not declare an "import" as a dependency', () {
+      d.file(path.join(appPath, 'lib', 'library.dart'), r'''
+        import 'package:not_declared/not_declared.dart';
+      ''').create();
+
+      expectValidationWarning(strictDeps);
+    });
+  });
+}
```

Comment:
I'm guessing you already parse URIs with `Uri.parse`, or similar. That should ensure that the URI is actually valid URIs.

Package URIs have a more restricted syntax that you might also want to check. To validate those, you can either reuse the code in [package:package_config](https://github.com/dart-lang/package_config/blob/master/lib/src/util.dart) or aim lower (that code is written for `.packages` files which are very specific, actual imports might want to be more permissive).

The base rules for package URIs is:
- Starts with `package:`. 
- Has no authority part.
- The path doesn't start with a `/`,
- but with a path segment containing at least one non-`.` character and no `:` or `%` characters,
- followed by a `/`.
- Has no query part (but that's probably not important).

You can check that at the string level with a RegExp:
```dart
var pchar = r"!$&-\-0-9;=@-Z_a-z~";
const pcharDot = r"!$&-.0-9;=@-Z_a-z~";
var packageUriRE = new RegExp("^package:[$pcharDot]*[$pchar]$[pcharDot]*/");
```
or on a parsed URI:
```dart
const pchar = r"!$&-\-0-9;=@-Z_a-z~";
const pcharDot = r"!$&-.0-9;=@-Z_a-z~";
final packagePathUriRE = new RegExp("^[$pcharDot]*[$pchar]$[pcharDot]*/");
bool isPackageUriValid(Uri uri) {
  return (uri.scheme == "package" &&
      !uri.hasAuthority &&
      uri.path.matches(packageUriPathRE) &&
      !uri.hasQuery);
}
```